contracts/LogarithmicBuckets.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: GNU AGPLv3
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import "./DoubleLinkedList.sol";
+
+library LogarithmicBuckets {
+    using DoubleLinkedList for DoubleLinkedList.List;
+
+    struct BucketList {
+        mapping(uint256 => DoubleLinkedList.List) lists; // All the accounts.
+        mapping(address => uint256) indexOf;
+        uint256 maxIndex;
+    }
+
+    /// CONSTANTS ///
+
+    uint256 private constant LOGBASE = 2;
+
+    /// ERRORS ///
+
+    /// @notice Thrown when the address is zero at insertion.
+    error AddressIsZero();
+
+    /// INTERNAL ///
+
+    /// @notice Updates an account in the `_heap`.
+    /// @dev Only call this function when `_id` is in the `_heap` with value `_formerValue` or when `_id` is not in the `_heap` with `_formerValue` equal to 0.
+    function update(
+        BucketList storage _buckets,
+        address _id,
+        uint256 _formerValue,
+        uint256 _newValue
+    ) internal {
+        uint96 formerValue = SafeCast.toUint96(_formerValue);
+        uint96 newValue = SafeCast.toUint96(_newValue);
+
+        if (formerValue != newValue) {
+            if (formerValue != 0) remove(_buckets, _id);
+            if (newValue != 0) insert(_buckets, _id, newValue);
+        }
+    }
+
+    /// PRIVATE ///
+
+    function insert(
+        BucketList storage _buckets,
+        address _id,
+        uint96 _value
+    ) internal {
+        // `_buckets` cannot contain the 0 address.
+        if (_id == address(0)) revert AddressIsZero();
+        uint256 bucketIndex = computeBucketIndex(_value);
+        _buckets.lists[bucketIndex].insertTail(_id, _value);
+        _buckets.indexOf[_id] = bucketIndex;
+        if (bucketIndex > _buckets.maxIndex) _buckets.maxIndex = bucketIndex;
+    }
+
+    /// @notice Removes an account in the `_buckets`.
+    /// @param _buckets The buckets to modify.
+    /// @param _id The address of the account to remove.
+    function remove(BucketList storage _buckets, address _id) internal {
+        uint256 index = _buckets.indexOf[_id];
+        // Swap the last account and the account to remove, then pop it.
+        _buckets.lists[index].remove(_id);
+        delete _buckets.indexOf[_id];
+
+        while (_buckets.lists[index].head == address(0) && index > 0) {
+            index -= 1;
+            _buckets.maxIndex = index;
+        }
+    }
+
+    /// @notice Compute the bucket index.
+    /// @param _value The value of the index to compute.
+    function computeBucketIndex(uint96 _value) private pure returns (uint256) {
+        for (uint256 i = 0; i < 128; i++) {
+            if (_value < LOGBASE**i) {
+                return i;
+            }
```

Comment:
I think it's more this to check no?
 
```suggestion
            if (LOGBASE**i <=_value  && _value < LOGBASE**i) {
                return i;
            }
```

Reply:
you mean i-1 no? but since the i's are increasing only 1 check is necessary.

Reply:
No `i` because you're starting from 0, no?

Reply:
You can't be both greater than and strictly lower than `LOGBASE**i`, so I don't think the code you suggest works @MerlinEgalite. Also, since we already checked that the value is not strictly smaller that a certain threshold in the previous iterations, I think it would be redundant to check it. In the end the code @Morphojo wrote seems to be working

Reply:
ah yes forgot the +1

Reply:
Yes it works my bad